java/c/src/main/java/org/apache/arrow/c/ReferenceCountedArrowArray.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.c;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.ForeignAllocationManager;
+
+/**
+ * The owner of an imported C Data Interface array.
+ *
+ * <p>There is a fundamental mismatch here between memory allocation schemes: AllocationManager represents a single
+ * allocation (= a single address and length). But an ArrowArray combines multiple allocations behind a single
+ * deallocation callback. This class bridges the two by tracking a reference count, so that the single callback
```

Comment:
> But an ArrowArray combines multiple allocations behind a single deallocation callback.

Wasn't the deallocation of an ArrowArray always follow the postorder-traversal of the array tree? Or maybe I just got something wrong about the deallocation process.

Reply:
The order does not matter here. It is about one vs many.

Reply:
Yes I got it after going through the code. Thought the comment is related to the struct array -> child array case.

Reply:
Ah ok yeah. I removed the original comment about the struct/child array case since it is just not possible to separate the child allocations from the parent, unfortunately.